File: src/week_8/ThirtyTwo.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.stream.*;

public class ThirtyTwo {

    private static List<String> stopwords = new ArrayList<>();

    private static List<String> partition(String data, int nLines) {
        List<String> lines = Arrays.asList(data.split("\n"));
        List<String> chunks = new ArrayList<>();
        for (int i = 0; i < lines.size(); i += nLines) {
            chunks.add(String.join("\n", lines.subList(i, Math.min(i + nLines, lines.size()))));
        }
        return chunks;
    }

    private static List<Map.Entry<String, Integer>> splitWords(String data) {
        String[] words = data.replaceAll("[\\W_]+", " ").toLowerCase().split("\\s+");
        List<Map.Entry<String, Integer>> result = new ArrayList<>();
        for (String w : words) {
            if (!stopwords.contains(w) && !w.isBlank()) {
                result.add(new AbstractMap.SimpleEntry<>(w, 1));
            }
        }
        return result;
    }

    private static Map<String, List<Map.Entry<String, Integer>>> regroup(Stream<List<Map.Entry<String, Integer>>> pairsList) {
        Map<String, List<Map.Entry<String, Integer>>> map = new HashMap<>();
        pairsList.forEach(list -> list.forEach(pair -> map.computeIfAbsent(pair.getKey(), k -> new ArrayList<>()).add(pair)));
        return map;
    }

    private static Map.Entry<String, Integer> countWords(Map.Entry<String, List<Map.Entry<String, Integer>>> entry) {
        int sum = entry.getValue().stream().map(Map.Entry::getValue).reduce(0, Integer::sum);
        return new AbstractMap.SimpleEntry<>(entry.getKey(), sum);
    }

    public static void main(String[] args) throws IOException {
        List<String> lines = Files.readAllLines(Paths.get("../stop_words.txt"));
        lines.forEach(l -> stopwords.addAll(List.of(l.split(","))));
        stopwords.addAll(Arrays.asList("a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z"));

        String data = Files.readString(Paths.get(args[0]));
        List<String> chunks = partition(data, 200);
        Stream<List<Map.Entry<String, Integer>>> splits = chunks.stream().map(ThirtyTwo::splitWords);
        Map<String, List<Map.Entry<String, Integer>>> groups = regroup(splits);
        List<Map.Entry<String, Integer>> wordFreqs = groups.entrySet().stream()
                .map(ThirtyTwo::countWords)
                .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
                .toList();

        wordFreqs.subList(0, 25).forEach(e -> System.out.println(e.getKey() + " - " + e.getValue()));
    }
}

File: src/week_8/ThirtyTwo.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"sort"
	"strings"
)

func main() {
	filePath := os.Args[1]
	stopWordsFile := "../stop_words.txt"

	stopWordsData, _ := ioutil.ReadFile(stopWordsFile)
	stopWords := strings.Split(string(stopWordsData), ",")

	data, _ := ioutil.ReadFile(filePath)
	dataStr := string(data)

	splits := mapSplitWords(partition(dataStr, 200), stopWords)
	splitsPerWord := regroup(splits)
	wordFreqs := sortWordFreqs(mapCountWords(splitsPerWord))

	for _, wf := range wordFreqs[:25] {
		fmt.Println(wf.word, "-", wf.count)
	}
}

type wordFreq32 struct {
	word  string
	count int
}

// generator like yield
func partition(dataStr string, nlines int) chan string {
	lines := strings.Split(dataStr, "\n")
	ch := make(chan string)
	go func() {
		for i := 0; i < len(lines); i += nlines {
			ch <- strings.Join(lines[i:min(i+nlines, len(lines))], "\n")
		}
		close(ch)
	}()
	return ch
}
func min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

func mapSplitWords(partitions chan string, stopWords []string) [][]wordFreq32 {
	var result [][]wordFreq32
	for p := range partitions {
		if p != "" {
			result = append(result, splitWords(p, stopWords))
		} else {
			break
		}
	}
	return result
}

func splitWords(dataStr string, stopWords []string) []wordFreq32 {
	wordRegexp := regexp.MustCompile(`[a-z]{2,}`)
	words := wordRegexp.FindAllString(strings.ToLower(dataStr), -1)
	var wordFreqs []wordFreq32
	for _, word := range words {
		if !contains32(stopWords, word) {
			wordFreqs = append(wordFreqs, wordFreq32{word, 1})
		}
	}
	return wordFreqs
}

func regroup(pairsList [][]wordFreq32) map[string][]wordFreq32 {
	mapping := make(map[string][]wordFreq32)
	for _, pairs := range pairsList {
		for _, p := range pairs {
			mapping[p.word] = append(mapping[p.word], p)
		}
	}
	return mapping
}

func mapCountWords(splitsPerWord map[string][]wordFreq32) []wordFreq32 {
	var wordFreqs []wordFreq32
	for _, v := range splitsPerWord {
		wordFreqs = append(wordFreqs, countWords(v))
	}
	return wordFreqs
}

func countWords(mapping []wordFreq32) wordFreq32 {
	count := 0
	for _, pair := range mapping {
		count += pair.count
	}
	return wordFreq32{mapping[0].word, count}
}

func sortWordFreqs(wordFreqs []wordFreq32) []wordFreq32 {
	sort.Slice(wordFreqs, func(i, j int) bool {
		return wordFreqs[i].count > wordFreqs[j].count
	})
	return wordFreqs
}

func contains32(arr []string, str string) bool {
	for _, a := range arr {
		if a == str {
			return true
		}
	}
	return false
}
